edgedb.go

```diff
@@ -74,36 +81,98 @@ func connectOne(ctx context.Context, cfg *connConfig, conn *baseConn) error {
 		err error
 	)
 
+	conn.writer = buff.NewWriter()
+
 	for _, addr := range cfg.addrs { // nolint:gocritic
-		// todo do error values need to be checked?
 		conn.conn, err = d.DialContext(ctx, addr.network, addr.address)
 		if err != nil {
-			log.Printf("while attempting connection %+v: %+v", addr, err)
+			err = wrapAll(err)
+			continue
+		}
+
+		toBeDeserialized := make(chan *soc.Data, 2)
+		r := buff.NewReader(toBeDeserialized)
+		go soc.Read(conn.conn, soc.NewMemPool(4, 256*1024), toBeDeserialized)
+
+		err = conn.setDeadline(ctx)
+		if err != nil {
+			err = wrapAll(err, conn.conn.Close())
 			continue
 		}
 
-		err = conn.connect(ctx, cfg)
+		err = conn.connect(r, cfg)
 		if err != nil {
-			_ = conn.conn.Close()
-			log.Printf("while attempting connection %+v: %+v", addr, err)
+			err = wrapAll(err, conn.conn.Close())
 			continue
 		}
 
+		conn.acquireReaderSignal = make(chan struct{}, 1)
```

Comment:
I'm wondering how creating a channel for `struct{}` is better than creating one for, say, booleans here? IMO sending only `struct{}{}` over it looks weird.

Reply:
Empty structs are semantic in go.
https://dave.cheney.net/2014/03/25/the-empty-struct
https://medium.com/@l.peppoloni/how-to-improve-your-go-code-with-empty-structs-3bd0c66bc531

In this case I see two advantages to using `struct{}{}` here.
1. `struct{}{}` has no value so it is clear that what comes out of the channel doesn't mean anything except that some thing came out of the channel. If it was `chan bool` one might mistakenly assign meaning to a `true` value that is somehow different than a `false` value.
2. `struct{}{}` has a width of `0` so it doesn't take any memory where `bool` has size of `1`. This difference is trivial in this case.

Reply:
Thanks for explaining! 🎓